### ISSUE TITLE
Use stock Deployments, DaemonSets, StatefulSets as pipelines.

### ISSUE THREAD

**Issue by abourget** [pachyderm/pachyderm#1611]
What would be awesome, is to be able to use stock `Deployments` or `StatefulSets` to provision Pachyderm pipelines.

The `pach-bin` could be explained and kept to an absolute minimum, so that we can leverage all the K8s features in our pipelines:
* PVs for temporary storage instead of the host's drive.
* Using secrets or config maps in ENV vars or mounts.
* Changing the Image Pull policy (I would have liked to use `:latest` all the way.. but I needed to delete, create or update pipelines)
* Deciding on securityContext for our container instead of having it always privileged.
* We could use StatefulSet for handling PersistentVolumeTemplates for parallel jobs..
* We could more easily use tags to select the nodes on which they should run:
  * 5 of my nodes are allows to access an internal database, per public IP.. I need certain jobs to run only there.
  * Some ML tasks will need special hardware, thus node affinity will be very useful.
* Would simplify the `pachyderm` code base.. we wouldn't need the `constant` or `coefficient` thing.. it'd be specified as we want on the cluster, either with an absolute number, a fixed number.. or through node affinity.

We could define layers on top of all of that.. but defining the contract between `pach-init` and Deployments, ReplicationCOntrollers and StatefulSets would be ssuuppperb.

**Comment by derekchiang** [comment]
Related: #1481

**Comment by derekchiang** [comment]
Discussed offline with @jdoliner.  We should implement a standard interface for processes to register themselves as pipelines.  That way you can create pipelines by any means (including `Deployments`).

**Comment by abourget** [comment]
I'm looking at the `args` passed to an `sh` and I can't find that information in the generated ReplicationController.  So it seems that `create-pipeline` stores some metadata _not_ in the RC. This would need to be revisited if we were to allow definition of pipelines solely by means of `kubectl create -f my-pipeline-deployment.yaml` ..

we could have `pachyderm.io/sh-arguments` as an annotation, and that would be picked up by the job dispatcher somehow ?! :)

**Comment by sjezewski** [comment]
Generally, we moved away from using k8s primitives to define pipelines because of the overhead it incurred. Spinning up a pod to process a datum was too slow. So we switched to a worker model.

That said, we'd love to use the k8s objects more directly. We'd love to have a pachyderm pipeline really just be a k8s manifest definition. (We were just discussing this again yesterday in the context of re-using other k8s components).

_That_ said, this is something we have an eye on, but we don't have a tactical plan to get there in the near term. So I'm closing this issue.
